### PR TITLE
Lua rewrite

### LINKED ISSUE #25: The post window has no memory
The post window does not "remember" what size it was when toggled back and forth with Enter.

I expected to be able to resize the post window, then toggle it off and on again retaining the resizing I had done.

Resize post window, toggle to off, toggle to on.

I think that this might be a useful feature as we are progressing more and more into look and feel of our environments being the equivalent to pyrotechnics at an AC/DC show, however subtle and understated, the devil is ever-present in the details.

NVIM v0.3.1
Mac OS 10.13.6
Running scnvim in Kitty version 0.12.3

### LINKED ISSUE #122: [FR]: Change SCNvimHelp / keywordprog / pandoc buffer position
I'd really like to open the help window in a vertical buffer, on the right side next to my code. Is that somehow possible?

I tried putting `setlocal keywordprg=:vert\ SCNvimHelp` in after/ftplugin/supercollider.vim, but it isn't working. I open my regular help files by having this in my init.vim:
`autocmd! FileType help wincmd L`
but could not find a working solution for the SuperCollider help.

When looking through your code, I found the following in scnvim/autload/help.vim (here on GitHub), in the function `scnvim#help#open`:
`execute 'topleft split | '`

Is this the place where something should be changed? I'm just not sure where to try that on my system, I installed SCNvim via a plugin manager (and would love to keep it updated that way). I'm a bit of a newbie with these things.

Any help (or tweak) would be appreciated!

EDIT: I think the most ideal solution would be to have it behave like the sclang post window so I can call it from different tabs - with the current solution, pressing K always brings me to the tab where help has been last opened.
I guess my ideal solution would be to have it behave like the post window and maybe even map it to Ctrl - Enter to toogle view.

### LINKED ISSUE #132: Replace internal calls to matching lua functions
- `scnvim#sclang`
- `scnvim#postwin`

### LINKED ISSUE #134: [FR]: Install as a system package
**Is your feature request related to a problem? Please describe.**
Hi! I package supercollider and a lot of other pro-audio tooling for Arch Linux. I'm a fan of vim and the scvim plugin, which is why I naturally also had a look at this plugin for neovim (and it's pretty awesome!).

I would like to be able to install this plugin as a system-wide neovim plugin using a (system) package manager (I know that a few other Linux distributions would be interested in that as well).

Ideally this would happen before tagging 1.0.0 (https://github.com/davidgranstrom/scnvim/issues/95).

**Describe the solution you'd like**
Install the plugin contents to a system-wide location.
Install the supercollider files to a system-wide location.
Create stateful content (the assets) in a per-user context (e.g. XDG compliant) and use it from those locations (this could even be a general solution for the plugin, to not generate untracked content in the plugin directory in a per-user installation).

**Describe alternatives you've considered**
Using a neovim/vim specific package manager can be an alternative, but I believe, that system package management allows for nicer OS level integration (and discoverability, e.g. via [optdepends](https://wiki.archlinux.org/index.php/PKGBUILD#optdepends)).

**Additional context**
I have created a [branch](https://github.com/dvzrv/scnvim/tree/install_system_wide) for a proof-of-concept for this (and I'm currently successfully using this with [this AUR package](https://aur.archlinux.org/packages/neovim-scnvim-git/)).
It features a Makefile (because that's well understood, simple, easy to maintain and available cross-platform) to install the files to their respective locations, which can be used when packaging this plugin (system-wide).

Additionally I have had to make two small changes, allowing to set a specific directory for stateful content (the assets):

https://github.com/davidgranstrom/scnvim/compare/master...dvzrv:install_system_wide

I'd be very happy about any comments specifically for the latter and how to integrate this further. At its current state the plugin now creates the required `scnvim-data` and `syntax` directories (on a Unix system) below a `g:scnvim_assets_dir` and uses the files from there, if `g:scnvim_assets_dir` (e.g. set to `~/.local/share/scnvim`) exists.

I think ideally the plugin would detect whether it is run from a system-wide location and based on that decide where to create the files and from where to source the files. The current approach is still a little hacky (although it works) and is a bit conflated with what `g:scnvim_root_dir` is supposed to be used for (AFAIK).

All in all I believe that making the plugin work in a system-wide context and a per-user context will not require a lot of changes or create a huge maintenance burden either. In any case: I'd be happy to help maintain this part of the code base.

### LINKED ISSUE #135: Move generated content to system specific locations
See discussion in #134

### LINKED ISSUE #136: Platform specific path handling
Centralize all system specific path handling (utils.vim etc.)

### LINKED ISSUE #147: [Question]: Evaluating when working with several buffers
**Describe the bug**

I found that when working with several files open in buffers, the ability to evaluate code blocks, surrounded by parenthesis, disappears except from the first file opened. 

Is there an init.vim setting for letting that behavior continue to be used across all open buffers in window? 

You can still evaluate by selecting the entire block in visual mode. 

**Expected behavior**

Still being able to evaluate big chunks of code without selecting all of it.

**Steps to reproduce**

Opening several buffers, rotating  between them and sending blocks of code to server.

**Additional context**

**Information**
- `nvim --version`:
- Operating system: mac os 11
- Running scnvim in a Terminal/GUI: terminal

### LINKED ISSUE #156: [BUG] scala file type overrides .sc filetype mapping
**Describe the bug**
When opening a .sc file the detected file type becomes scala. This is not a problem with .scd files. Not entirely sure whether this is a bug or a missing feature in the README.

**Expected behavior**
I would like to be able to see in the README how to efficiently override this nvim behaviour, and be able to set .sc files to be openend as filetype supercollider.

**Steps to reproduce**
Open .sc file with nvim.

**Additional context**
The filetype value is set through an autocmd from augroup filetypedetect in `/usr/share/nvim/runtime/filetype.vim`. This loads the `/usr/share/nvim/runtime/ftplugin/scala.vim`, which for instance overrides the indentation settings for scnvim (which is why I discovered this behaviour)

**Information**
- `nvim --version`: v0.6.0-dev+635-ga31e7523f
- Operating system: Arch Linux  5.15.2-arch1-1
- Running scnvim in a Terminal/GUI: terminal

### LINKED ISSUE #183: Upcoming breaking changes
Hi all,

This is a tracking issue where I will post breaking changes in the upcoming version of scnvim. 
If you want to try it out you can follow along in #153. It starting to become pretty stable by now but changes might still occur. I will try and post here if I break something intentionally.

Overall goals of this refactor/rewrite have been to rewrite a lot of viml code to lua to ease maintainability and coding standards (and in some cases improve performance). Fix some long standing issues in the code base that has been hard to solve because of legacy code and implement some new cool features (floating post window I'm looking at you!). I hope that the new configuration system will make it easier to start using scnvim out-of-the-box and to be able to configure the system to your liking.

I will write a more definite guide on how to transition to the new lua configuration once the API has settled.

## Configuration

- All global options have been removed (e.g. `g:scnvim_postwin_direction` etc.) these are instead declared in `scnvim.setup`
- All `<Plug>` mappings have been removed, mappings can now be defined directly in the setup function
  - No mappings are defined by default
  - It's possible to create your own mappings using the scnvim API if you don't like to define them in the setup function

Please see the [default configuration](https://github.com/davidgranstrom/scnvim/blob/topic/lua-config/lua/scnvim/config.lua) for what options are available. **The rewrite is still in motion, so things might change without notice.**

Example for setting options with the new configuration:

```lua
local scnvim = require 'scnvim'
scnvim.setup {
   mapping = {
    ['<M-e>'] = scnvim.map.send_line({'i', 'n'}),
    ['<C-e>'] = {
      scnvim.map.send_block({'i', 'n'}),
      scnvim.map.send_selection('x'),
    },
    ['<F12>'] = scnvim.map.hard_stop({'n', 'x', 'i'}),
    ['<CR>'] = scnvim.map.postwin_toggle('n'),
    ['<M-CR>'] = scnvim.map.postwin_toggle('i'),
    ['<M-L>'] = scnvim.map.postwin_clear({'n', 'i'}),
    ['<C-k>'] = scnvim.map.show_signature({'n', 'i'}),
  },
  editor = {
    flash = {
      duration = 100,
      repeats = 2,
      hl_group = 'TermCursor',
    },
  },
}
```

## Custom mappings

For custom mappings that relied on the viml api `scnvim#sclang#send` you can instead use `scnvim.send`

```vim
" old
nnoremap <leader>sm :call scnvim#sclang#send_silent('s.meter')<cr>
" new
nnoremap <leader>sm <cmd>lua require('scnvim').send_silent('s.meter')<cr>
```

## API

- The viml API is gone (e.g. `scnvim#sclang#send_silent`) and have been replaced by equivalent lua functions
- Some functions have moved to avoid cyclic references between modules
  - `utils.path_sep` => `path.sep`
  - `utils.get_scnvim_root_dir` => `path.get_plugin_root_dir`

### PR DESCRIPTION
**Pull request by davidgranstrom**
This PR is now a rewrite of the scnvim project. It contains some breaking changes, notably the viml API is gone and is now implemented in lua.

**New features**

* New mapping system
  * Be able to pre-process text before being sent to sclang
* Post window can be displayed as a floating window
* "fade" eval highlight effect
* (Regular) Post window remember its last size
* More fine grained user configuration (and hopefully easier)
* scnvim (SuperCollider) classes will be installed automatically when calling `scnvim.setup`
* Bug fixes (some important)
* Extensions

**Developer features**

* Code base formatted with `stylua`
* Auto generated and hosted API documentation
* Unit tests
* CI (lint/style, unit tests)
* Extensions API

**Breaking changes**

* Neovim version 0.7 or higher
* No public viml API with the exception of `scnvim#statusline#server_status` which still is convenient for configuring the status line.
* `SCNvimTags` has been renamed to `SCNvimGenerateAssets`
* No default mappings are defined (use the `scnvim.setup` function instead)

See [`lua/scnvim/config.lua`](https://github.com/davidgranstrom/scnvim/blob/topic/lua-config/lua/scnvim/config.lua) for default configuration values.

View the auto generated API documentation [here](https://davidgranstrom.github.io/scnvim/)

Example configuration:

```lua
local scnvim = require 'scnvim'
local map = scnvim.map
local map_expr = scnvim.map_expr

scnvim.setup {
  keymaps = {
    ['<M-e>'] = map('editor.send_line', {'i', 'n'}),
    ['<C-e>'] = {
      map('editor.send_block', {'i', 'n'}),
      map('editor.send_selection', 'x'),
    },
    ['<CR>'] = map('postwin.toggle'),
    ['<M-CR>'] = map('postwin.toggle', 'i'),
    ['<M-L>'] = map('postwin.clear', {'n', 'i'}),
    ['<C-k>'] = map('signature.show', {'n', 'i'}),
    ['<F12>'] = map('sclang.hard_stop', {'n', 'x', 'i'}),
    ['<leader>st'] = map('sclang.start'),
    ['<leader>sk'] = map('sclang.recompile'),
    ['<F1>'] = map_expr('s.boot'),
    ['<F2>'] = map_expr('s.meter'),
  },
  editor = {
    highlight = {
      color = 'IncSearch',
      type = 'flash',
    },
  },
  documentation = {
    cmd = '/opt/homebrew/bin/pandoc',
  },
  postwin = {
    keymaps = {
      q = map('postwin.close')
    },
    float = {
      enabled = true,
    },
  },
}
```

Close #136 
Close #25 
Close #132 
Close #135 
Close #136 
Close #134 (the generated assets aspect of the feature request)
Close #147 
Close #122 
Close #156
Close #183

### CODE REVIEW

**Comment by davidgranstrom** [comment]
Did an experiment today with the new mapping system. It lets you hook a callback function to be executed before the text is actually sent to SuperCollider. Could be used for text substitution (kind of like a preprocessor), but also to call any other arbitrary function on a specific scnvim key action/mapping. 

By default the mapping table is empty, but I will add a example config with all the current default mappings for easy transition to the new lua config.

```lua
local scnvim = require'scnvim'

scnvim.setup {
  mapping = {
    -- Map 'send_line' in insert and normal mode
    ['<C-e>'] = scnvim.map({'i', 'n'}, 'send_line'),

    -- Map 'send_line' in insert and normal mode and filter the line before sending.
    ['<C-e>'] = scnvim.map({'i', 'n'}, 'send_line', function(data)
      local line = data[1]
      line = line:gsub('goodbye', 'hello')
      return {line}
    end),
  },
}
```

**Comment by davidgranstrom** [comment]
@madskjeldgaard Glad to hear! :)

Yes, you can use `offset_x` and `offset_y` to position the float in the config, here's all options that you can use: https://github.com/davidgranstrom/scnvim/blob/topic/lua-config/lua/scnvim/config.lua#L37-L50

**Comment by davidgranstrom** [comment]
..thinking about this now, I think I might add `row` and `col` options (as functions) in order to specify the absolute position from the config.

File: autoload/scnvim/editor.vim [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    endif
+    call setpos('.', [0, start_pos[0], start_pos[1], 0])
+    let end_pos = s:find_match('(', ')', forward_flags)
+    " restore cursor
+    call setpos('.', c_curpos)
+    return [start_pos[0], end_pos[0]]
+endfunction
+
+" Could be replaced with `vim.region` later
+function! scnvim#editor#get_visual_selection() abort
```
Comment (ranjithshegde) [inline autoload/scnvim/editor.vim:48]: Just an FYI, if you are keeping these purely for the lack of get_visual_range in lua, I use this in my rc
```lua
function utils.get_visual_selection(bufnr)
    vim.api.nvim_input "<Esc>"
    vim.api.nvim_input "gv"
    local first = vim.api.nvim_buf_get_mark(bufnr, "<")
    local last = vim.api.nvim_buf_get_mark(bufnr, ">")
    local lines = vim.api.nvim_buf_get_lines(bufnr, first[1], last[1], true)

    return {
        lines = lines,
        line_start = first[1],
        line_end = last[1],
        col_start = first[2],
        col_end = last[2],
    }
end
```

Comment (davidgranstrom) [reply autoload/scnvim/editor.vim:48]: (sorry, my previous comment was apparently not saved before marking this as resolved)

Thanks! These two functions will be converted to lua eventually, I'd just like to get a few more unit test for the editor module before doing the refactoring. I'm planning to do a follow up PR for this after this one has been is merged.

**Comment by madskjeldgaard** [comment]
I'm wondering something: How does one map something like a custom piece of sc code using the new mapping paradigm? Eg. `Server.local.boot` ? 

Also:
Maybe it's worth having recompile as a mappable function in `scnvim/editor.lua` ? (And maybe also Server.local.boot ? And potentially SCNvimStart)

**Comment by madskjeldgaard** [comment]
As an example of what I imagine a mappable piece of sc code would like like in the new paradigm:

```
	['<F1>'] = scnvim.map.execute_sclang("Server.local.boot", {'n', 'x', 'i'})
```

**Comment by davidgranstrom** [comment]
Hi @madskjeldgaard,

It is possible to map any arbitrary function using the `map` helper, here are some examples:

```lua
    ['<leader>st'] = map(scnvim.start),
    ['<leader>sk'] = map(scnvim.recompile),
    ['<leader>sh'] = map(function()
      scnvim.send('"hello!".postln')
    end)
```

Your example could be written as:

```lua
    ['<F1>'] = scnvim.map(function()
      scnvim.send("Server.local.boot")
    end, {'n', 'x', 'i'}),
```

If you plan to do a lot of these, perhaps this wrapper function could serve as a starting point to make the mappings a bit more clear (maybe I'll add this to the docs too):

```lua
local map_expr = function(expr, modes)
  modes = modes or {'n', 'x', 'i'}
  return scnvim.map(function()
    scnvim.send(expr)
  end, modes)
end

-- usage
scnvim.setup {
  mapping = {
    ['<F1>'] = map_expr("Server.local.boot"),
    ['<F2>'] = map_expr("s.meter"),
    }
}
```

**Comment by madskjeldgaard** [comment]
> ```lua
> , {'n', 'x', 'i'}
> ```

fantastic!

**Comment by madskjeldgaard** [comment]
As far as I can tell the docs don't contain info on changing post window size/position (yet), just FYI

**Comment by madskjeldgaard** [comment]
Also, this is just a small thing possibly for later - but how could one have these new mappings be used in other filetypes other than SuperCollider? I'm thinking for example of Faust files where having access to post window etc. would be cool.

**Comment by davidgranstrom** [comment]
Yes, (user) documentation is still lacking in some aspects. There is an overview in `:h scnvim`, but it's not comprehensive. 

I have not decided exactly on how to proceed to create a more detailed user documentation.. the API documentation is auto generated (from the lua comments) so it's always is up-to-date:

https://davidgranstrom.github.io/scnvim/modules/scnvim.config.html#default.postwin

(There is also a page for the `config` module.)

The default values are set here:

https://github.com/davidgranstrom/scnvim/blob/topic/lua-config/lua/scnvim/config.lua

**Comment by davidgranstrom** [comment]
> Also, this is just a small thing possibly for later - but how could one have these new mappings be used in other filetypes other than SuperCollider? I'm thinking for example of Faust files where having access to post window etc. would be cool.

The mappings are applied for the `supercollider` filetype only (via an autocmd). But we could expose the local `apply_keymaps` function in the editor module and then it would be possible to apply keymaps for other filetypes as well, do you think that would be sufficient? 

If it's only for a specific function (like toggle the post window), then using `lua.keymap.set` and calling the scnvim API could also be an alternative:

```lua
vim.keymap.set('n', '<leader>st', '<cmd>lua require"scnvim.postwin".toggle()<cr>')
```

(It is possible to not specify any mappings at all in the config and use any nvim mapping method instead)

**Comment by madskjeldgaard** [comment]
> https://davidgranstrom.github.io/scnvim/modules/scnvim.config.html#default.postwin

this is super nice!!!

**Comment by madskjeldgaard** [comment]
More random things:
postwin width and height must be integers otherwise it results in errors. if Using eg. `vim.fn.winwidth(0) / 2` in a function one might easily get a float. It's easily fixed with math.floor(), but maybe that should be done in the function that receives this value?

**Comment by madskjeldgaard** [comment]
A possible bug: It seems that when choosing Luasnip as snippet engine, the produced snippet is not properly escaped with it's `\` characters. Here is an example from the snippets produced on my system:

```
s( {trig = "FoscStartTextSpan.new", name = "FoscStartTextSpan.new", dscr = "Snippet for FoscStartTextSpan.new, auto generated by SCNvim" }, {t("FoscStartTextSpan.new"),t("("),t("leftText:"),i(1, "nil"),t(", "), t("rightText:"),i(2, "nil"),t(", "), t("style:"),i(3, "nil"),t(", "), t("direction:"),i(4, "nil"),t(", "), t("command:"),i(5, "\startTextSpan"),t(", "), t("concatHspaceLeft:"),i(6, "0.5"),t(", "), t("concatHspaceRight:"),i(7, "0.5"),t(", "), t("leftBrokenText:"),i(8, "nil"),t(", "), t("rightPadding:"),i(9, "nil"),t(", "), t("tweaks:"),i(10, "nil"), t(")"),}),
```

**Comment by madskjeldgaard** [comment]
Also it might be worthwile to include the `local scnvim = require("scnvim")` line in the help file example for setting up scnvim using the new paradigm. This implicit require might confuse new users a bit :)

**Comment by madskjeldgaard** [comment]
Another thing missing from the docs: The help file should include an example mapping for looking up help file for word under cursor, as I think that one might be a bit complex for some people.

**Comment by davidgranstrom** [comment]
@madskjeldgaard Great notes about the docs, I'll add your suggestions.

**Comment by davidgranstrom** [comment]
I think this is ready to be merged now.

**Comment by madskjeldgaard** [comment]
> I think this is ready to be merged now.

I agree. Do it !

**Comment by madskjeldgaard** [comment]
> A possible bug: It seems that when choosing Luasnip as snippet engine, the produced snippet is not properly escaped with it's `\` characters. Here is an example from the snippets produced on my system:
> 
> ```lua
> s( {trig = "FoscStartTextSpan.new", name = "FoscStartTextSpan.new", dscr = "Snippet for FoscStartTextSpan.new, auto generated by SCNvim" }, {t("FoscStartTextSpan.new"),t("("),t("leftText:"),i(1, "nil"),t(", "), t("rightText:"),i(2, "nil"),t(", "), t("style:"),i(3, "nil"),t(", "), t("direction:"),i(4, "nil"),t(", "), t("command:"),i(5, "\startTextSpan"),t(", "), t("concatHspaceLeft:"),i(6, "0.5"),t(", "), t("concatHspaceRight:"),i(7, "0.5"),t(", "), t("leftBrokenText:"),i(8, "nil"),t(", "), t("rightPadding:"),i(9, "nil"),t(", "), t("tweaks:"),i(10, "nil"), t(")"),}),
> ```

This is the only issue at the moment, but I'm not sure if it's related to this. Maybe more relevant in a seperate PR. What do you think David?

**Comment by davidgranstrom** [comment]
> > A possible bug: It seems that when choosing Luasnip as snippet engine, the produced snippet is not properly escaped with it's `\` characters. Here is an example from the snippets produced on my system:
> > ```lua
> > s( {trig = "FoscStartTextSpan.new", name = "FoscStartTextSpan.new", dscr = "Snippet for FoscStartTextSpan.new, auto generated by SCNvim" }, {t("FoscStartTextSpan.new"),t("("),t("leftText:"),i(1, "nil"),t(", "), t("rightText:"),i(2, "nil"),t(", "), t("style:"),i(3, "nil"),t(", "), t("direction:"),i(4, "nil"),t(", "), t("command:"),i(5, "\startTextSpan"),t(", "), t("concatHspaceLeft:"),i(6, "0.5"),t(", "), t("concatHspaceRight:"),i(7, "0.5"),t(", "), t("leftBrokenText:"),i(8, "nil"),t(", "), t("rightPadding:"),i(9, "nil"),t(", "), t("tweaks:"),i(10, "nil"), t(")"),}),
> > ```
> 
> This is the only issue at the moment, but I'm not sure if it's related to this. Maybe more relevant in a seperate PR. What do you think David?

I'm not so familiar with the luasnip snippet generator code, but I could have a quick look later today. It's probably an escaping issue for the `\` character. We could make a follow up PR for that separately otherwise so not to delay merging this any further. Will do some minor fixes for the README tonight and then we should be good to go!

**Comment by madskjeldgaard** [comment]
One last idea: Add an option to have the key bindings be mapped to the post window as well as the help browser window. I personally use this and perhaps others might find this useful?

**Comment by davidgranstrom** [comment]
> One last idea: Add an option to have the key bindings be mapped to the post window as well as the help browser window. I personally use this and perhaps others might find this useful?

Ah, forgot about the help window, definitely useful to be able to run code examples :) I think I'll add an option in the `postwin` and `documentation` tables to easily toggle this on/off.

**Comment by davidgranstrom** [comment]
> One last idea: Add an option to have the key bindings be mapped to the post window as well as the help browser window. I personally use this and perhaps others might find this useful?

Good catch again. This should now work with the latest push, both `postwin` and `documentation`. `documentation` has the same mappings by default, postwin has none.

```lua
  postwin = {
    mapping = { -- or set the true to use the same values as top-level `mapping`
      q = map(require('scnvim.postwin').close)
    },
}
```

I'm having some second thoughts about the mapping system.. I'm not so happy with the code duplication in `editor` needed for the current mapping approach. I have a solution I would like to try before the merge, better now than later :)

**Comment by davidgranstrom** [comment]
Here's the idea for the new `map` helper:

```lua
  mappings = {
    ['<M-e>'] = map('editor.send_line', {'i', 'n'}),
    ['<leader>st'] = map(scnvim.start),
    ['<F1>'] = map_expr('s.boot'),
    ['<F2>'] = map_expr('s.meter'),
  },
```

The `map` helper object can take a string or a function. If the argument is a string, it will parse the module name and the function name (e.g. `editor.send_line`) and create a mapping based on that. There is some validation to what modules can be used, mainly for debugging purposes (spelling mistakes etc.)
 
The `map_expr` helper is an addition to make it easy to map any valid SuperCollider expression to a key.

**Comment by davidgranstrom** [comment]
Breaking changes for keymaps: https://github.com/davidgranstrom/scnvim/issues/183#issuecomment-1170939934

**Pull request merged**